components/clp-package-utils/clp_package_utils/controller.py

```diff
@@ -0,0 +1,783 @@
+        :param updates: The updates.
+        :raise ValueError: If a key in `updates` doesn't exist in `settings`.
+        """
+        for key, value in updates.items():
+            if key not in settings:
+                error_msg = (
+                    f"{parent_key_prefix}{key} is not a valid configuration key for the webui."
+                )
+                raise ValueError(error_msg)
+            if isinstance(value, dict):
+                self._update_settings_object(f"{parent_key_prefix}{key}.", settings[key], value)
+            else:
+                settings[key] = value
+
+
+class DockerComposeController(BaseController):
+    """
+    Controller for orchestrating CLP components using Docker Compose.
+    """
+
+    def __init__(self, clp_config: CLPConfig, instance_id: str) -> None:
+        self._project_name = f"clp-package-{instance_id}"
+        super().__init__(clp_config)
+
+    def start(self) -> None:
+        """
+        Starts CLP's components using Docker Compose.
+
+        :raise: Propagates `subprocess.run`'s exceptions.
+        """
+        check_docker_dependencies(
+            should_compose_project_be_running=False, project_name=self._project_name
+        )
+        self._set_up_env()
+
+        deployment_type = self._clp_config.get_deployment_type()
+        logger.info(f"Starting CLP using Docker Compose ({deployment_type} deployment)...")
+
+        cmd = ["docker", "compose", "--project-name", self._project_name]
+        if deployment_type == DeploymentType.BASE:
+            cmd += ["--file", "docker-compose.base.yaml"]
+        cmd += ["up", "--detach", "--wait"]
+        subprocess.run(
+            cmd,
+            cwd=self._clp_home,
+            check=True,
+        )
+        logger.info("CLP is started.")
+
+    def stop(self) -> None:
+        """
+        Stops CLP components deployed via Docker Compose.
+
+        :raise: Propagates `subprocess.run`'s exceptions.
+        """
+        try:
+            check_docker_dependencies(
+                should_compose_project_be_running=True, project_name=self._project_name
+            )
+        except OSError as e:
+            logger.warning(
+                'Docker dependencies check failed: "%s". Attempting to stop CLP containers '
+                "anyway...",
+                e,
+            )
+        else:
+            logger.info("Stopping all CLP containers using Docker Compose...")
+
+        subprocess.run(
+            ["docker", "compose", "--project-name", self._project_name, "down"],
+            cwd=self._clp_home,
+            check=True,
+        )
+        logger.info("All CLP containers stopped.")
+
+    @staticmethod
+    def _get_num_workers() -> int:
+        """
+        TODO: Revisit after moving from single-container to multi-container workers. See issue
+        @y-scope/clp#1424 for details.
+        :return: Number of worker processes to run.
+        """
+        return multiprocessing.cpu_count() // 2
```

Comment:
```suggestion
        # This will change when we move from single to multi-container workers. See y-scope/clp#1424
        # for details.
        return multiprocessing.cpu_count() // 2
```